clio-client/src/main/scala/org/broadinstitute/clio/client/metadata/CramUndeliverer.scala

```diff
@@ -0,0 +1,105 @@
+package org.broadinstitute.clio.client.metadata
+
+import java.net.URI
+
+import better.files.File
+import org.broadinstitute.clio.client.dispatch.MoveExecutor.{IoOp, MoveOp, WriteOp}
+import org.broadinstitute.clio.transfer.model.cram.{CramExtensions, CramMetadata}
+
+class CramUndeliverer extends MetadataMover[CramMetadata] {
+  override protected def moveMetadata(
+    src: CramMetadata,
+    destination: URI,
+    newBasename: Option[String]
+  ): (CramMetadata, Iterable[IoOp]) = {
+    val movedCram = src.cramPath.map(
+      MetadataMover
+        .buildFilePath(_, destination, newBasename.map(_ + CramExtensions.CramExtension))
+    )
+
+    val movedCramsDest = src.copy(
+      cramPath = movedCram,
+      craiPath = movedCram.map(
+        cramUri => URI.create(s"$cramUri${CramExtensions.CraiExtensionAddition}")
+      )
+    )
+
+    val writeMd5Op = for {
+      md5 <- src.cramMd5
+      cram <- movedCram
+    } yield {
+      WriteOp(md5.name, URI.create(s"$cram${CramExtensions.Md5ExtensionAddition}"))
+    }
+
+    val ops = Iterable(
+      extractMoves(src, movedCramsDest, _.cramPath),
+      extractMoves(src, movedCramsDest, _.craiPath),
+      writeMd5Op.toIterable
+    ).flatten
+
+    val undeliverSampleMetrics: Boolean = src.cramPath
+      .map(_.getHost)
+      .exists(
+        cramHost => src.alignmentSummaryMetricsPath.exists(_.getHost.equals(cramHost))
+      )
+
+    if (undeliverSampleMetrics) {
+      if (src.regulatoryDesignation.exists(_.isClinical)) {
+        throw new RuntimeException(
+          s"Cannot undeliver sample level metrics for clinical sample $src"
+        )
+      }
+      lazy val oldBaseName =
+        src.cramPath.map(
+          p => File(p.getPath).name.replace(CramExtensions.CramExtension, "")
+        )
+
+      def makeDestMetrics(srcMetric: Option[URI]): Option[URI] = {
+        srcMetric.map { metric =>
+          val srcMetricFileName = File(metric.getPath).name
+          val destMetricFileName = (for {
+            nbn <- newBasename
+            obn <- oldBaseName
+          } yield {
+            srcMetricFileName.replace(obn, nbn)
+          }).getOrElse(srcMetricFileName)
+          destination.resolve(destMetricFileName)
+        }
+      }
+
+      val moveMetricsOps =
+        src.sampleLevelMetrics.flatten
+          .zip(src.sampleLevelMetrics.flatMap(makeDestMetrics))
+          .map(MoveOp.tupled)
+
+      val movedMetricsDest = movedCramsDest.copy(
```

Comment:
I didn’t check, but I assume that this copies a lot of the code in deliver, since the same paths are affected. Is there any way to avoid the code duplication?